### ISSUE TITLE
Negative loss values when doing off-policy evaluation

### ISSUE THREAD

**Issue by xinofekuator** [VowpalWabbit/vowpal_wabbit#3151]
#### Describe the bug
I get negative average loss after doing off-policy evaluation with`dr` method. This tutorial was followed: https://vowpalwabbit.org/tutorials/off_policy_evaluation.html 

#### To Reproduce
1. Create train.dat with the following content:
```
1:1:0.5 | user_age:25
2:0:0.5 | user_age:25
2:1:0.5 | user_age:55
2:1:0.5 | user_age:56
2:1:0.5 | user_age:55
2:1:0.5 | user_age:56
1:1:0.5 | user_age:27
2:0:0.5 | user_age:21
2:0:0.5 | user_age:23
2:0:0.5 | user_age:56
2:1:0.5 | user_age:55
2:1:0.5 | user_age:36
1:1:0.5 | user_age:25
2:0:0.5 | user_age:25
2:1:0.5 | user_age:55
2:1:0.5 | user_age:56
```
2. `vw --cb 2 --cb_type dr -d train.dat -f candidate-model.vw --save_resume  --passes 10 --cache_file /tmp/train.cache`

When I run this I get this output:
```
final_regressor = candidate-model.vw
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
using cache_file = /tmp/train.cache
ignoring text input in favor of cache input
num sources = 1
Enabled reductions: gd, scorer, csoaa, cb
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
2.000000 2.000000            1            1.0    known        1        2
1.000000 0.000000            2            2.0    known        2        2
1.340543 1.681085            4            4.0    known        2        2
1.008415 0.676288            8            8.0    known        2        2
-0.461373 -0.461373           16           16.0    known        2        2 h
-0.544761 -0.628150           32           32.0    known        2        2 h
-0.638947 -0.733132           64           64.0    known        2        2 h
-0.737488 -0.836029          128          128.0    known        2        2 h

finished run
number of examples per pass = 15
passes used = 10
weighted example sum = 150.000000
weighted label sum = 0.000000
average loss = -0.901187 h
total feature number = 300
```

The loss computed in the holdout data (10% of the provided data) is negative. I thought average loss in vw was supposed to be positive and not sure how to interpret this.

The same result can be obtained with `vw --cb 2 -d train.dat -f candidate-model.vw --save_resume  --passes 10 --cache_file /tmp/train.cache` as doubly robust is the default one.

Changing the default squared loss to classic one also yields negative results: `vw --cb 2 -d train.dat -f candidate-model.vw --save_resume  --passes 10 --cache_file /tmp/train.cache --loss_function classic`

#### Expected behavior
Positive average loss

#### Observed Behavior
Negative average loss

#### Environment
MacOS Catalina with vw version `8.10.1 (git commit: 3887696)`

**Comment by ataymano** [comment]
It is known open issue (that should be fixed) about DR loss being reported when --cb_type dr: #2222
Let us review and re-prioritize it next week.
But while reported loss is DR, there is no specific reason for it to be positive.

**Comment by jackgerrits** [comment]
@ataymano is this also related? https://github.com/VowpalWabbit/vowpal_wabbit/issues/1696

**Comment by ataymano** [comment]
looks like #2222 as action item created as followup for #1696?

**Comment by jackgerrits** [comment]
If they all are referring to the same thing then we should close others as duplicates?

**Comment by ataymano** [comment]
yeah and probably #2222 should be the one that is open as the most actionable one.

**Comment by ataymano** [comment]
This one is technically slightly different, so ideally we should double check if problem reproduces after #2222 is closed. Not sure if we should close it right now.

**Comment by xinofekuator** [comment]
> But while reported loss is DR, there is no specific reason for it to be positive.

Is *the lower the better* the correct way of interpreting the DR loss?

The idea I had was to run a grid search only with DR method so I do not mind using DR loss for picking the best hyperparams. But I am not sure about what is the correct way to interpret these results.

**Comment by xinofekuator** [comment]
One additional question: how does the DR loss interacts with the `loss_function` argument? 
When providing classic and squared results are different, but I still do not understand how the squared loss equation can be negative. The only explanation I find is that that loss function is not really being used.

**Comment by xinofekuator** [comment]
One difference I see with https://github.com/VowpalWabbit/vowpal_wabbit/issues/1696:

@JohnLangford wrote:
> This seems to not be a bug. If you use --noconstant, the average loss changes to zero. What's happening is that the parameter on the constant feature is being adjusted towards the observed cost (-1) resulting in the second prediction having a nonzero value according to the double robust methodology.

I check this by using `--noconstant` but it does not make the average loss positive: `vw --cb 2 -d train.dat -f candidate-model.vw --save_resume  --passes 10 --cache_file /tmp/train.cache --noconstant` and all costs are positive in this training data so not sure why the constant would be negative.

**Comment by xinofekuator** [comment]
> This one is technically slightly different

@ataymano Based on my previous comment what it was discussed in the other issue does not apply here, so maybe it is a different issue.

**Comment by ataymano** [comment]
@xinofekuator - sorry, missed your previous comment. Looking.

**Comment by ataymano** [comment]
Sorry for delay in response.

I think that comment was very specific for the dataset discussed in #1696 and not relevant much to the rest of discussion (the problem of reported average loss are not being determined by predictions).

I have double checked that the problem is still about the same thing:
loss reported as average loss is calculated as DR one here:
https://github.com/VowpalWabbit/vowpal_wabbit/blob/6d7193a0ff440af55403356bfacbd4f45df7bbd1/vowpalwabbit/cb_algs.h#L68
and there is no specific reason for DR loss to be within convex hull of reported losses.

Slightly simplified way to reproduce:
Data: 

2:1:0.5 | user_age:56
1:1:0.5 | user_age:27
2:0:0.5 | user_age:21


Command line:
`
vw -d data.txt --cb 2 -k --cb_type dr --holdout_after 2 -P 1
`

I did quick and dirty test (replacing dr-cost calculation referenced above with ips one in place) and was not able to get negative losses (as expected).

At the same time, although there is an open item about separating learning algorithm and estimator, [both DR and IPS are unbiased](https://github.com/VowpalWabbit/vowpal_wabbit/issues/1696#issuecomment-449989608) so you can compare them to each other (in the way you described - smaller is better)

**Comment by ataymano** [comment]
Looks like there is nothing wrong with negative loss reported in this case. @xinofekuator, please let me know if you still have questions or we can close this issue.

**Comment by ataymano** [comment]
Closing this issue, but please let me know if re-opening is required.
